src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -108,38 +105,48 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     		}
     	} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
-			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
-				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
-				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+			Locale locale = Locale.forLanguageTag(dto.getLocale());
+			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
+				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
+					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
+			} else { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(ps.getSupportedLocales(), locale);
+				if (matchedLocale != null) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
 					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
-					return new ComponentService(matchedLocaleDTO).getMessages();
-				}
+					cacheItem = new ComponentService(matchedLocaleDTO).getMessages();
+				} else  // Requested locale is not supported and does not match any supported locales
+					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 			}
-
-			//Create and store cacheItem for the requested locale
-			cacheItem = createCacheItem(fallbackLocalesIter);
     	}
     	return cacheItem;
     }
 
+    /**
+     * Gets the messages in the next fallback locale by passing the next fallback locale DTO to a new instance of ComponentService
+	 * and then invoking {@link #getMessages(Iterator)}.
+	 * @param fallbackLocalesIter The fallback locale queue to use in case of failure. If null, no locale fallback will be applied.
+     */
+    private MessageCacheItem getFallbackLocaleMessages(Iterator<Locale> fallbackLocalesIter) {
+		if (!dto.getLocale().equals(ConstantsKeys.SOURCE) && fallbackLocalesIter != null && fallbackLocalesIter.hasNext()) {
```

Comment:
Is it necessary to judge dto locale is same as the fallback locale?

Reply:
It's not necessary, but it is an improvement. If we do not check for dto.getLocale == fallback and it happens, it will just try to fetch for the dto locale again and will fail again, and move on to the next item in the queue. Nonetheless, I have added the check for optimization.

Moreover, I hsve removed the following restriction to allow falling back to SOURCE messages:
`!dto.getLocale().equals(ConstantsKeys.SOURCE)`